Add basic syslog implementation

crest42 opened the pull request:
### Contribution description
For a university project, I added a small and lightweight syslog implementation for RIOT. The implementation included in this pull request is a working prototype with a few features missing (see below). I added this pull request to get a general review if this feature is desired by the RIOT community and if so, to get feedback about the design of my implementation.

@miri64 You mentioned interest in such an implementation in #2376 back in 2016. I hope this is still relevant :)

The implementation consists of two parts:

1. The user interface

   This Interface is inspired by the glibc implementation of syslog 
   (https://www.gnu.org/software/libc/manual/html_node/Overview-of-Syslog.html) and should enable a user to easily commit messages to a local syslogd instance using generic RIOT IPC.

   The implementation is really simple. When a user logs a string, a syslog msg is built and wrapped into a syslog_msg_t. A pointer to this msg is transferred via RIOT IPC.

2. The syslogd implementation

   To get a lightweight implementation that is easily extensible with RIOT IPC and that uses no async I/O and stuff like that, I wrote a very basic implementation of a syslogd, which works on a single input and multiplexes syslog messages on this input to multiple outputs. To not blew up the RIOT codebase i implemented this syslogd as a pkg.

   Missing Features:
   - Outputs are very basic by now 
   - Documentation
   - syslog.conf support
   - Network Input

   Supported Features:
   - Logging to stdout
   - Logging to a remote syslog server (tested with a local installation of syslog-ng)
   - Logging to file (Test still missing)
   - The implementation should be rfc5424 compatible

### Testing procedure

Atm only native is tested. I plan to make a further test with a nucleo-f767zi in the future.

1. cd tests/syslog
2. Update the ipv6 address of the remote logging target if remote logging is needed
3. make all term
4. syslog_init
5. syslog "msg" or syslog_test

Expected Output:
```
> syslog_init
syslog_init
> syslog "msg"
syslog "msg"
<8> 1 - barkon.chilio.net test 0 0 msg
> syslog_test
syslog_test
<11> 1 - example.com test 0 0 Logging Logging 123 3
<14> 1 - example.com test 0 0 Logging Logging 123 6
<8> 1 - example.com test 0 0 TestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTestTest
<8> 1 - example.com test 0 0 
<9> 1 - example.com test 0 0 
<10> 1 - example.com test 0 0 
<11> 1 - example.com test 0 0 
<12> 1 - example.com test 0 0 
<13> 1 - example.com test 0 0 
<14> 1 - example.com test 0 0 
> 
```

### Issues/PRs references
resolve #2376.

makomi commented on sys/syslog/syslog.c:
```diff
@@ -0,0 +1,80 @@
+static char _ident[IDENT_MAX_LEN];
+static char hostname[HOST_NAME_MAX];
+static mutex_t syslog_mutex = MUTEX_INIT;
+static int _facility = 0;
```
If you like, you can drop the explicit initialization of  `_facility`, since it will be initialized to `0` anyway. See sections 6.2.4 (3) and 6.7.8 (10) of a [draft of the ISO C (ANSI C) standard](http://www.open-std.org/JTC1/SC22/WG14/www/docs/n1256.pdf).

crest42 replied:
Yea that is true. But it should be auto-optimized by the compiler anyway? Or is it just always good practice to drop such redundant initializations? I asking bc I find it a little bit more readable with an explicit initialization.

makomi replied:
True, the compiler does not care whether one writes the additional assignment.

I would drop it, since for me it makes the code easier to read but I am not sure if it is considered best practice and I am not aware of any related RIOT-OS guideline. One might argue that omitting the explicit initialization hides what is happening, making the code more difficult to read.

I guess, I just wanted to point it out and see what you think of it, since you explicitly initialized `_facility` but not e.g. `_ident`. But to be clear, I don't think it really matters. This was just my curiosity speaking.

crest42 replied:
Ok, I was curious if there are any implications I am not aware of. Anyways, it should be at least consistent in my code. I will have a look into it (Maybe zero isn't even the best idea). Thanks for your feedback!

crest42 commented on sys/shell/commands/syslog.c:
```diff
@@ -0,0 +1,37 @@
+    if(argc < 2) {
+        return -1;
+    }
+    syslog(LOG_EMERG, argv[1], LOG_ERR);
```
Remove LOG_ERR

crest42 commented:
Implemented piped message queue backend as discussed with @kaspar030  and @jcarrano 

Added the following tests:
- tests/log: Default log backend used by RIOT
- tests/log_mqueue: Message queue backend. Overwrites LOG() functionality by writing to a pipe. The pipe is read by a thread and messages are dispatched according to their facility value
- tests/log_mqueue_syslog: Add syslog functionality, which overwrites dispatch targets of log_mqueue. LOG() calls with a facility != 0 gets dispatched to syslog implementation
- tests/syslog: Uses syslog without special backend for compatibility. Calls to LOG() gets treated as usual

crest42 commented:
The implementation needs a little bit of rework. Obvious problems right now are:

- ~~Priority values in log.h and priority values in syslog.h are not the same by now. This could lead to Messages being accepted by LOG() but discarded by syslog(). Maybe i could change the implementation provided by log.h without losing backwards compatibility.~~
- The dispatch logic needs further rethinking. By now an auto_init function plainly overwrites the whole dispatcher array with facilities != 0. Also a linked list may be more feasible to allow more than a single module to attach onto a single facility value.
- The syslog implementation hooks itself to log_mqueue as soon as the log_mqueue module is used. This may should be more configurable.
- Configurability of queue sizes could be better.
- Missing Documentation.
